Refactor ImageNet example aligning with Issue #231

mohitreddy1996 opened the pull request:
Refactor Imagenet example aligning with #231 

Major changes:
- Refactor files to example_lib (containing training and eval loop code), example_main (script which parses flags and starts the training loop), model_name (containing necessary model definition), input_pipeline (bucketing the entire preprocessing code).

- Introduce a test using tfds.testing.mock_data which executes a single step of training and eval.
  - In order to do so, copy the metadata files from TFDS and store in examples/testing/datasets.
  - Document necessary changes required to be made by other examples.

avital commented on examples/imagenet/input_pipeline.py:
```diff
@@ -57,13 +56,21 @@ def distorted_bounding_box_crop(image_bytes,
     max_attempts: An optional `int`. Number of attempts at generating a cropped
         region of the image of the specified constraints. After `max_attempts`
         failures, return the entire image.
+    rng: RNG used for generating seeds for reproducibility of
```
Instead of this, add a comment that we need to resolve reproducibility issues, linking to a GitHub issue with details

mohitreddy1996 replied:
Created the issue and mentioned it in the README file.

avital commented on examples/imagenet/imagenet_benchmark.py:
```diff
@@ -41,14 +32,12 @@ class ImagenetBenchmark(Benchmark):
 
     start_time = time.time()
-    train.main([])
+    imagenet_lib.train_and_evaluate(model_dir=model_dir, batch_size=2048,
```
I originally suggested to @AlexeyG to do the benchmarks this way but he persuaded me that it would be good to have the benchmarks exactly line up with the arguments you'd pass in the command line to the main train python file.

What do you think?

mohitreddy1996 replied:
Most of the cases, just set a few flag. 
Worst case, same as passing them to the example_lib.method().

I think I like calling the main idea better, Thanks for noticing this!

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -44,56 +31,37 @@
 
 
-def create_model(key, batch_size, image_size, model_dtype):
+def _create_model(key, batch_size, image_size, model_dtype):
```
Let's keep these public methods.

mohitreddy1996 replied:
Done. 

Since we wanted to have type annotation for public methods with appropriate docstrings, I felt that the ones which are local/private, we could stick with the standard.

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -191,58 +159,78 @@ def _prepare(x):
-  it = map(prepare_tf_data, ds)
+def _create_input_iter(dataset_builder, batch_size, image_size, dtype, train,
+                       cache):
+  ds = input_pipeline.create_split(
```
Can you tell me more about this TFDS API change? What are "dataset builders"?

mohitreddy1996 replied:
The reason for this was majorly because to get rid of previously globally declared vars:
TRAIN_IMAGES = 1281167
EVAL_IMAGES = 50000

To have an extendable solution, we shouldn't have hardcoded values of something which is controlled by a different library. In order to fetch this information using the TDFS library itself, I modified the existing code to use DatasetBuilder (an abstract base class for every dataset). DatasetBuilder contains information around dataset such as: feature names, types, shapes, version, splits etc.

Also the dataset_builder is initialized in the main train loop to avoid multiple initializations which increases the chances of inconsistent changes.

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -191,58 +159,78 @@ def _prepare(x):
 # transformations like tree_map and pmap.
 @flax.struct.dataclass
-class TrainState:
+class _TrainState:
```
ditto on this and all the others, we don't need to add underscores to any of these

mohitreddy1996 replied:
done.

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -191,58 +159,78 @@ def _prepare(x):
+      should be written to.
+    batch_size: Batch size of the input.
+    num_epochs: Number of epochs to cycle through before stopping.
+    learning_rate: Learning rate for the momentum optimizer.
```
This isn't accurate. This argument is "the learning rate in case you have batch size 256. The effective learning rate is scaled linearly to the batch size"

mohitreddy1996 replied:
oh no... Thanks for catching (will make the necessary changes in #289 as well)

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -191,58 +159,78 @@ def _prepare(x):
+    cache: Determines whether the dataset should be cached.
+    half_precision: Determines whether bfloat16/float16 should be used
+      instead of float32.
+    num_train_and_eval_steps: Number of steps for training and eval.
```
Hmm... Do we need this? It's a bit off to see an argument designed just for a test. Are there any other ways to achieve this?

avital replied:
(I wrote another comment on this below where you call this function from your test)

mohitreddy1996 replied:
Although specifying the number of examples for train and test/eval in tfds.testing.mockdata, the base info such as num_examples in a split (train/eval) is same as the original dataset. It is documented as WARNING in [documentation](https://www.tensorflow.org/datasets/api_docs/python/tfds/testing/mock_data).

Without this flag, we would do a lot more than a single iteration :/ I couldn't think of something better.. 

We can discuss further in the below comment if refactoring is the way to go.

marcvanzee replied:
The WARNING in the documentation says that `info.split['train'].num_examples` doesn't match `len(list(ds_train))`. So how about you use `len(list(ds_train))` below instead of `.num_examples` to determine the size of the dataset? In that case I think you won't need this argument.

avital commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -191,58 +159,78 @@ def _prepare(x):
+      instead of float32.
+    num_train_and_eval_steps: Number of steps for training and eval.
+      This is used for testing (default: -1 i.e use the entire dataset).
+    disable_checkpointing: Determines whether the training and evaluation
```
I'm also not sure about this. Our official examples need to be as short as possible. What's wrong with always saving checkpoints? We should also test that...

mohitreddy1996 replied:
So I had seen multiple instances where checkpointing took quite a bit of wall time in test execution during development (although I ran the test multiple times now, no such issues :/). So I decided to disable it, given how `checkpoints` is a different library and should have testing of its own. 

But I guess we still need to ensure that after the last (and only) step, checkpoints are saved in tests. Removed this flag.

avital commented on examples/imagenet/imagenet_lib_test.py:
```diff
@@ -0,0 +1,84 @@
+def _parse_and_return_summary_values(path):
+  """Parses event file in given `path` and returns scalar summaries logged."""
+  tag_event_value_dict = {}
+  event_file_generator = directory_watcher.DirectoryWatcher(
```
Why do we need this? Are we really watching the directory? Isn't there a more straightforward API to just read the latest observations? (What did Alexey do in his original Benchmark PR?)

mohitreddy1996 replied:
This is required to test the recorded metrics as an assertion that the model training indeed happened.
So this code is a motivation from that PR. This is mostly copy pasted from there and I have #290 to address this duplication.

mohitreddy1996 replied:
Removed the assertions altogether.

avital commented on examples/imagenet/imagenet_lib_test.py:
```diff
@@ -0,0 +1,84 @@
+  return tag_event_value_dict
+
+
+class ImageNetLibTest(absltest.TestCase):
```
Just call it `ImageNetTest`

mohitreddy1996 replied:
done.

avital commented on examples/imagenet/imagenet_lib_test.py:
```diff
@@ -0,0 +1,84 @@
+      imagenet_lib.train_and_evaluate(
+          model_dir=model_dir, batch_size=8, num_epochs=1,
+          learning_rate=0.1, momentum=0.9, cache=False, half_precision=False,
+          num_train_and_eval_steps=1, disable_checkpointing=True)
```
I see that this is the place that you need `num_train_and_eval_steps=1`. Instead, we should consider refactoring `train_and_evaluate` into smaller functions and then just call into those directly instead of having this override. That would also clarify the different parts of the ImageNet train code.

avital commented on examples/imagenet/imagenet_lib_test.py:
```diff
@@ -0,0 +1,84 @@
+
+    summary_values_dict = _parse_and_return_summary_values(path=model_dir)
+
+    # Since the values could change due to stochasticity in input processing
```
If we're not actually looking at the numbers at all there's no point in these assertions. I don't mind saying that this only tests that nothing breaks after one step of training, and the other full end-to-end tests test the meaningful part -- that training gets to the expected numbers.

mohitreddy1996 replied:
Hmm, that's true. Without reproducibility of the example, having these assertions are meaningless :/ 

Although we could just put values like loss greater than 6.0 or something. I believe the accuracy would stay at 0.0 since the input used is always the same and the remaining randomness (is in the input preprocessing logic with stateful TF Image ops). 

WDYT? I am TBH inclining towards removing it. Just a thought.

marcvanzee replied:
Remove them makes sense to me.

mohitreddy1996 replied:
Thanks! Removed it.

avital commented on examples/imagenet/input_pipeline.py:
```diff
@@ -21,22 +21,18 @@
-                                aspect_ratio_range=(0.75, 1.33),
-                                area_range=(0.05, 1.0),
-                                max_attempts=100):
+def _distorted_bounding_box_crop(image_bytes,
```
No underscores before function names

mohitreddy1996 replied:
done.

avital commented on examples/imagenet/input_pipeline.py:
```diff
@@ -126,21 +122,21 @@ def _decode_and_center_crop(image_bytes, image_size):
-  crop_window = tf.stack([offset_height, offset_width,
-                          padded_center_crop_size, padded_center_crop_size])
-  image = tf.image.decode_and_crop_jpeg(image_bytes, crop_window, channels=3)
+  crop_window = tf.stack([offset_height, offset_width, padded_center_crop_size,
```
nit but please don't change whitespace unless there's a reason -- it makes merge conflicts unnecessarily likely

mohitreddy1996 replied:
reverted the changes.

avital commented on examples/imagenet/input_pipeline.py:
```diff
@@ -172,42 +168,50 @@ def preprocess_for_eval(image_bytes, dtype=tf.float32, image_size=IMAGE_SIZE):
   """Creates a split from the ImageNet dataset using TensorFlow Datasets.
 
   Args:
+    dataset_builder: TFDS dataset builder for ImageNet.
```
Can't the dataset builder just be constructed here rather than at the top level of the training loop?

mohitreddy1996 replied:
Discussed in the previous comment. Basically to avoid multiple initializations of the dataset_builder. 

The values of number of train and eval images are required in main train loop to compute the steps per epoch and here as well to compute split. We definitely don't want to have hardcoded numbers. So initialized it in the main train loop and passing it on as argument.

avital commented on examples/imagenet/input_pipeline.py:
```diff
@@ -216,9 +220,11 @@ def decode_example(example):
   if train:
     ds = ds.repeat()
-    ds = ds.shuffle(16 * batch_size, seed=0)
+    ds = ds.shuffle(16 * batch_size)
```
Why get rid of the seed here?

mohitreddy1996 replied:
sorry my bad. I was trying to introduce the seeds using jax.random.split(seed) for reproducibility. Removed it assuming I introduced it. 

Thanks for catching it.

avital commented on examples/imagenet/resnet_v1.py:
```diff
@@ -13,18 +13,17 @@
-
 
-class ResidualBlock(nn.Module):
+class _ResidualBlock(nn.Module):
```
No underscores

mohitreddy1996 replied:
done.

avital commented on examples/imagenet/resnet_v1.py:
```diff
@@ -13,18 +13,17 @@
+class _ResidualBlock(nn.Module):
   """Bottleneck ResNet block."""
 
+  # pylint: disable=arguments-differ
```
Oh? What's this?

mohitreddy1996 replied:
Same as below. Removed.

avital commented on examples/imagenet/resnet_v1.py:
```diff
@@ -51,10 +50,25 @@ def apply(self, x, filters, strides=(1, 1), train=True, dtype=jnp.float32):
-  """ResNetV1."""
+  """ResNet V1."""
+
+  # pylint: disable=arguments-differ
```
Oh?

mohitreddy1996 replied:
aah my bad, I have a linter in VS code and it was complaining about it. I guess I will remove it.

avital commented on examples/testing/datasets/imagenet2012/5.0.0/image.image.json:
```diff
@@ -0,0 +1 @@
+{"encoding_format": "jpeg", "shape": [-1, -1, 3]}
```
Have you tried to see if we can avoid this by loading these files directly via

```
import tfds
tfds.__path__
```

mohitreddy1996 replied:
Yup, I gave that a shot. testing/metadata not exposed as part of the library in the python packages sadly. 

I spoke with the TFDS team, they recommended to copy the metadata files for now but a better solution is WIP (ETA: later this year).

avital commented on examples/testing/datasets/README.md:
```diff
@@ -0,0 +1,34 @@
+
+Example test: `flax/examples/imagenet_lib_test.py` 
+
+PR introducing the above mentioned changes: [PR](https://github.com/google/flax/pull/287).
```
I don't think this should be here. (And in general people can always `git blame`)

mohitreddy1996 replied:
Yup that makes sense.

avital commented:
@jheek As the "owner" of the ImageNet example, I'd like to get your review on this as well.

marcvanzee commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -123,7 +91,7 @@ def step_fn(step):
 def train_step(state, batch, learning_rate_fn):
   """Perform a single training step."""
-  def loss_fn(model):
+  def _loss_fn(model):
```
I think up to now we haven't used underscores for functions in our codebase, so I would avoid it here as well for consistency.

avital replied:
Or more generally, a pull request should only change things that pertain to that particular pull request.

If we think we should change our naming convention, we can discuss that but we shouldn't combine that with a pull request refactoring an example for sake of testing.

mohitreddy1996 replied:
I used this convention as this file had a similar naming convention (def _prepare(x)) in `prepare_tf_data`. Changed it back and also changed _prepare -> prepare. 

@avital yeah, I think that makes sense. Since this was a huge refactoring, I just put them the smaller ones in as well, probably wasn't a good idea :)

mohitreddy1996 commented on examples/imagenet/README.md:
```diff
@@ -18,7 +18,10 @@ The model should run with other configurations and hardware, but explicitely tes
+`python imagenet_main.py --batch_size=2048 --cache=True --model_dir=./imagenet_fp16_bs2048 --half_precision=True --loss_scaling=256.`
+
+### Reproducibility
+Making the ImageNet classification example reproducible in WIP. For more details, follow [#291](https://github.com/google/flax/issues/291).
```
Thanks!

mohitreddy1996 commented:
@marcvanzee 

> Re "The WARNING in the documentation says that info.split['train'].num_examples doesn't match len(list(ds_train)). So how about you use len(list(ds_train)) below instead of .num_examples to determine the size of the dataset? In that case I think you won't need this argument."

Please correct me if I am wrong, but `info.split['train'].num_examples` is basically reading a field which I believe is present in tfds metadata. But when we do len(list(ds_train)), do you think we might be doing additional transformations? The main reason for introducing a `test-only` field was primarily that. WDYT? @avital what are your thoughts on this?

Marvin182 commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "editor.tabSize": 2
```
Do you want this to be part of this CL?

mohitreddy1996 replied:
My bad, this was accidentally added as one of the commit. This should be an outdated file now.

Marvin182 commented on examples/imagenet/imagenet_lib.py:
```diff
@@ -44,42 +32,23 @@
 import jax.numpy as jnp
 
 import tensorflow.compat.v2 as tf
```
/tensorflow.compat.v2/tensorflow

Marvin182 commented on examples/imagenet/imagenet_lib_test.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class ImageNetLibTest(absltest.TestCase):
+
```
You might want to add:
```
  def setUp(self):
    super().setUp()
    tf.config.experimental.set_visible_devices([], "GPU")
```
This avoids some OOM error when the test case runs on GPUs.

mohitreddy1996 replied:
Great suggestion, thanks!

Marvin182 commented on examples/imagenet/imagenet_main.py:
```diff
@@ -0,0 +1,73 @@
+from absl import app
+from absl import flags
+
+import tensorflow.compat.v2 as tf
```
/tensorflow.compat.v2/tensorflow

mohitreddy1996 replied:
Again outdated file :) 

Replaced it in all instances of imagenet example.

Marvin182 commented on examples/imagenet/input_pipeline_test.py:
```diff
@@ -0,0 +1,54 @@
+
+  def test_train_and_evaluate(self):
+    # Get testing directory.
+    tf.enable_v2_behavior()
```
not needed. I think we don't support TF1 in Flax.

mohitreddy1996 replied:
Done! Thanks for pointing this out.

Marvin182 reviewed (approved):
Looks good. We could do more changes (like switching to ml_collections.ConfigDict for the hyperparameters), but let's do it in separate PRs and in coordination with Andreas.
